### PR TITLE
feat(connlib): Add `file_logger` module and configure via `connect`

### PR DESCRIPTION
**Pull request by jamilbk**
Implements the FFI part of the following issues:

Fixes firezone/product#669
Refs firezone/firezone#2018
Refs firezone/firezone#2017

~~Log wg stats as well by moving into warn category.~~ This contains information we want to keep out of our logging infrastructure for now.

### CODE REVIEW

File: kotlin/android/app/src/main/java/dev/firezone/android/tunnel/TunnelManager.kt
Comment (jamilbk) [inline kotlin/android/app/src/main/java/dev/firezone/android/tunnel/TunnelManager.kt:179]: @pratikvelani This seemed to somehow been reverted from #1949

File: rust/connlib/clients/android/Cargo.toml
Comment (jamilbk) [inline rust/connlib/clients/android/Cargo.toml:16]: @conectado This seemed to somehow been reverted from #1950

File: rust/connlib/clients/headless/src/main.rs
Comment (jamilbk) [inline rust/connlib/clients/headless/src/main.rs]: Do we actually want this somewhere in `$HOME/.config` or something?

**Review by roop — APPROVED** [review]
The Apple part looks good. In the future (after we enable app groups), the log dir shall be changed to a place that both the tunnel and the app can access.

**Comment by pratikvelani** [comment]
@jamilbk  I have a few queries on logging.

- When the debug configuration is supplied to the `Session.connect()` method, it presents an issue wherein users will need to disconnect and reconnect the session in order to activate debugging mode, toggling the switch will have no effect when the session is already up. In this scenario, it's crucial to inform users that debugging will only become effective upon restarting the session.
- Should we use the `Logger` class to pass the debug parameters? like have a static method (similar to `Logger.init()`) and pass the debug parameters to it?
- While logging should we also be tagging the logs with device info (like `Android/iOS version, manufacture, model info` also any other `device/network` info that could help reproduce the issue), Also do we need to pass `account-id`?
- Should we clean up the `log` folder when the user logs out?
- Is it feasible for the native lib to detect 'low storage' conditions to clear old logs?

File: rust/connlib/libs/common/src/session.rs
Comment (conectado) [inline rust/connlib/libs/common/src/session.rs]:
```suggestion
        log_dir: &Path,
```

With this we let the type system do the path sanity check :)

Comment (conectado) [reply rust/connlib/libs/common/src/session.rs]: If we still need an owned type use [PathBuf](https://doc.rust-lang.org/std/path/struct.PathBuf.html)

File: rust/connlib/libs/common/src/session.rs
Comment (conectado) [inline rust/connlib/libs/common/src/session.rs]: `tracing_subscriber` should be configured in the binaries not in the libraries, this will grant us more flexibility and reduce a bug if the binaries configure it thsemselves.

Furthermore, I think this will consume the logs and prevent us from seeing anything in stdout so we need to do something like I did [here](https://github.com/firezone/firezone-gateway/blob/main/gateway/src/logging.rs#L37-L63) with the old gateway.

**Review by conectado — CHANGES_REQUESTED** [review]

**Comment by jamilbk** [comment]
> @jamilbk I have a few queries on logging.

Good questions! Thanks for clarifying. I don't think we need to persist Kotlin-side UI logs, so this will be implemented on the native side so we can have the same behavior across all platforms. So Firebase for the app-level backtraces and native-side connlib for connectivity logging, I think that should suffice for v1.

> 
> * When the debug configuration is supplied to the `Session.connect()` method, it presents an issue wherein users will need to disconnect and reconnect the session in order to activate debugging mode, toggling the switch will have no effect when the session is already up. In this scenario, it's crucial to inform users that debugging will only become effective upon restarting the session.

Good catch. I forgot to note this in the Wireframe. I think it's OK to prompt the user "Changing debug mode will sign out your current session. Would you like to proceed?" when the user enables/disables debug mode.

> * Should we use the `Logger` class to pass the debug parameters? like have a static method (similar to `Logger.init()`) and pass the debug parameters to it?

Hmm. I'd prefer (I think) to keep this in the `connect` call so that's it's uniform across platforms. Android is the only one with a Logger init which is unique to Logcat.

> * While logging should we also be tagging the logs with device info (like `Android/iOS version, manufacture, model info` also any other `device/network` info that could help reproduce the issue), Also do we need to pass `account-id`?

Yeah, we're logging most of this in connlib already -- I don't think there's anything more on the Kotlin side we need for now.  `account-id` can be looked up in connlib already.

> * Should we clean up the `log` folder when the user logs out?

We _should_ be able to handle this on the native side.

> * Is it feasible for the native lib to detect 'low storage' conditions to clear old logs?

I think we can detect this on the native side and handle appropriately as well.

**Comment by jamilbk** [comment]
> The Apple part looks good. In the future (after we enable app groups), the log dir shall be changed to a place that both the tunnel and the app can access.

Hm, I think this was my intent with this line:

```
let cachesDirectory = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask)
```

Just for my own learning, does `FileManager` not handle that for us?

**Comment by pratikvelani** [comment]
Sounds great 👍, just one point

> > * Should we clean up the `log` folder when the user logs out?
> 
> We _should_ be able to handle this on the native side.

For the logout currently we are not sending any events to connlib. As the user should be able to independently logout from the `Settings` screen without the session being active or the connlib loaded, hence if required we can implement the clearing of log folder on kotlin side.

**Comment by jamilbk** [comment]
> For the logout currently we are not sending any events to connlib. As the user should be able to independently logout from the `Settings` screen without the session being active or the connlib loaded, hence if required we can implement the clearing of log folder on kotlin side.

Hmm signout and session disconnect should be tied together, i.e. when the user clicks "sign out" (or changes debug mode) that should call the connlib disconnect function which will stop logging.

There shouldn't be a case where the user is signed in but connlib is disconnected (at least for any extended period). If the token is invalid, connlib will detect and eventually call `onDisconnect`, which should be used to clear the token and move to signed out state.

Is it possible to store the logs in the app cache directory? That way the user can clear them from system settings. I don't know if we should expose log modification from the kotlin side because we will be uploading and deleting these logs from the native side, leading to possible race conditions.

File: kotlin/android/app/src/main/java/dev/firezone/android/tunnel/TunnelManager.kt
Comment (pratikvelani) [inline kotlin/android/app/src/main/java/dev/firezone/android/tunnel/TunnelManager.kt:158]:
```suggestion
                    logDir = appContext.cacheDir.absolutePath,
```

Logs can be stored in cache. `cacheDir` returns the reference to app's internal cache dir.

**Review by pratikvelani — APPROVED** [review]

File: rust/connlib/libs/common/src/session.rs
Comment (jamilbk) [reply rust/connlib/libs/common/src/session.rs]: Gotcha -- yeah it looks like `tracing` abstracts this away for us by expecting a `String` for the appender. So we may not need it for the `file_logger`. We'll probably need for the `log_uploader` module when I implemented that.

File: rust/connlib/libs/common/src/session.rs
Comment (jamilbk) [reply rust/connlib/libs/common/src/session.rs]: Hmm, gotcha. I've updated the PR to move all init logging logic into the clients' entrypoints (`main.rs` or `lib.rs`). From there we can set up `tracing-subscriber` appropriately, and then the rest of the `Session.connect` flow need not concern itself with this stuff.

I still need to split off the logs to STDOUT and the file like you mentioned above.

**Comment by jamilbk** [comment]
@conectado Did some refactoring to move the logging init into a `file_logger.rs` module. Also, I removed this from the Gateway -- I think that'll require a different logging approach for a couple reasons:

- The Gateway is in a more sensitive environment security-wise, and we don't want to store state or leave logs there
- There's a non-trivial likelihood the Gateway won't have outbound connectivity to GCP, so uploading logs will be blocked
- The Gateway (and self-hosted Relay) will be in predictable connectivity environments, so there's less value to be gleaned from these kinds of detailed connectivity logs.

^ The same points apply to the Relay. For these reasons, I believe we should keep this logging approach limited to the clients.

I can mark this as a draft PR until I set up the tracing subscribers as you mentioned @conectado. Will probably get to that tonight or early tomorrow.

After that, what's left TODO:

- [ ] Implement `file_uploader.rs` module which uses a connected ControlPlane to request an Upload URL and upload new file(s) to it periodically.
- [ ] firezone/firezone#2019
- [ ] firezone/firezone#2018
- [ ] firezone/firezone#2017

File: kotlin/android/app/src/main/java/dev/firezone/android/tunnel/TunnelLogger.kt
Comment (jamilbk) [inline kotlin/android/app/src/main/java/dev/firezone/android/tunnel/TunnelLogger.kt:1]: @pratikvelani As it turns out, we can just init the Android logger in the connect call like the other clients 🙃 

So I've removed this.

File: kotlin/android/app/src/main/java/dev/firezone/android/tunnel/TunnelManager.kt
Comment (jamilbk) [inline kotlin/android/app/src/main/java/dev/firezone/android/tunnel/TunnelManager.kt:204]: We didn't actually need a separate function for this. I've removed it in favor of an `init_logging` function in the native-side `connect` entrypoint

**Comment by roop** [comment]
> > The Apple part looks good. In the future (after we enable app groups), the log dir shall be 
> > changed to a place that both the tunnel and the app can access.

> Hm, I think this was my intent with this line:

> `let cachesDirectory = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask)`

> Just for my own learning, does FileManager not handle that for us?

The cachesDirectory for the app and the tunnel will be different. The app cannot access the tunnel's directory, and the tunnel cannot access the app's directory. For a shared disk location, we need to setup app groups and use [`containerURL(forSecurityApplicationGroupIdentifier:)`](https://developer.apple.com/documentation/foundation/filemanager/1412643-containerurl) -- we can use `Library/Caches` under that.

Let's decide whether we want to change it to the shared location or not in the future after we discuss the logging idea further.

File: rust/connlib/libs/common/src/session.rs
Comment (jamilbk) [reply rust/connlib/libs/common/src/session.rs]: Updated to use PathBuf

File: rust/connlib/libs/common/src/session.rs
Comment (jamilbk) [reply rust/connlib/libs/common/src/session.rs]: This has been updated to follow the approach you have in the old gateway using a `tracing_subscriber::registry()` with multiple layers to write to file and still log to the OS-appropriate log system.

**Comment by jamilbk** [comment]
Converting to draft to work through some issues not caught in the build.

**Comment by jamilbk** [comment]
K, fixed the crash on Android. Still getting an occasional crash on Apple ~~due to the File logger so going to bypass it for now.~~ Hm, scratch that. Not able to reproduce. cc @thomaseizinger

File: rust/connlib/libs/client/src/file_logger.rs
Comment (thomaseizinger) [inline rust/connlib/libs/client/src/file_logger.rs:14]: Any plans on how we are going to achieve that? :)

File: rust/connlib/clients/android/src/lib.rs
Comment (thomaseizinger) [inline rust/connlib/clients/android/src/lib.rs]: FYI: This will panic if called more than once. Not sure what the above comment refers to.

**Review by thomaseizinger — APPROVED** [review]
Rust code looks good!

File: rust/connlib/clients/android/src/lib.rs
Comment (jamilbk) [reply rust/connlib/clients/android/src/lib.rs]: Ah yes, I'll need to use `try_init` instead

File: rust/connlib/libs/client/src/file_logger.rs
Comment (jamilbk) [reply rust/connlib/libs/client/src/file_logger.rs:14]: Other than some rudimentary log filters, not at the moment. I suppose this comment was more as a reminder to myself and any future maintainers to remember not to explicitly log (at WARN and above) any of the above. Open to ideas / suggestions though!

**Comment by jamilbk** [comment]
Sample logfile entry (note this is in debug build, so INFO is logged):

```json
{"time":"2023-09-11T13:49:47.570624Z","target":"connlib","logging.googleapis.com/sourceLocation":{"file":"connlib/clients/apple/src/lib.rs","line":"149"},"severity":"INFO","message":"File logging initialized! Logging to \"/Users/jamil/Library/Containers/dev.firezone.firezone.network-extension/Data/Library/Caches/log\""}
```

File: rust/Cargo.toml
Comment (jamilbk) [inline rust/Cargo.toml:38]: @conectado -- looks like ring with its included cc-rs is building fine now, so I've left this out to avoid future issues. Wanted to keep it noted here because ring tends to have issues and it could pop back up.

**Review by AndrewDryga — APPROVED** [review]
CI part LGTM

**Comment by jamilbk** [comment]
So I'm going to go ahead and merge this in order to unblock other PRs, but @thomaseizinger it looks like the `tracing_appender` may be having an issue flushing logs to disk. The `WorkerGuard` that we capture with the `_guard` variable in the `init_logging` function in the client libs may be going out of scope, triggering `Drop` which will prevent the `tracing_appender` from flushing in a timely manner. The [docs advise](https://docs.rs/tracing-appender/latest/tracing_appender/non_blocking/struct.WorkerGuard.html) to create that variable at the entry point of the program, but since the clients use a lib approach, I'm not sure how best to address it.

That may not turn out to be the issue, but at the moment only the `INFO` log within `file_logger.rs` that I gave an example of is being printed.

File: rust/connlib/libs/client/src/file_logger.rs
Comment (conectado) [reply rust/connlib/libs/client/src/file_logger.rs:14]: I think we should never store that sensitive information directly in a string but create a wrapper `Sensitive<T>` that doesn't implement neither `Debug` nor `Display` and has a method `get_sensitive`, to reduce the chances of this happening.

Or maybe just using [`secrecy`](https://docs.rs/secrecy/latest/secrecy/index.html)'s `SecretString` is even better, I was already planning on using it for tokens.

File: rust/connlib/libs/client/src/file_logger.rs
Comment (conectado) [inline rust/connlib/libs/client/src/file_logger.rs:32]: Will this ever be printed? Since you're probably using this when creating the subscriber I'm not sure.

File: rust/connlib/libs/client/src/file_logger.rs
Comment (conectado) [inline rust/connlib/libs/client/src/file_logger.rs:45]: I think we should have a more customizable level-filter, this function could accept something that [`impl Filter`](https://docs.rs/tracing-subscriber/latest/tracing_subscriber/layer/trait.Filter.html)

Comment (conectado) [reply rust/connlib/libs/client/src/file_logger.rs:45]: This way we can more easily re-use the file logger.

**Review by conectado — APPROVED** [review]
LGTM! left some non-blocking comments.

File: rust/connlib/libs/client/src/file_logger.rs
Comment (jamilbk) [reply rust/connlib/libs/client/src/file_logger.rs:45]: Hmm, good point. How do you feel about leaving this out of the file logger entirely and letting the calling code apply the filter? E.g. in the `init_logging` function in the clients.

File: rust/connlib/libs/client/src/file_logger.rs
Comment (jamilbk) [reply rust/connlib/libs/client/src/file_logger.rs:32]: It depends on whether the client has another logging mechanism or not set up. On Android, since `android_logger` is configured before tracing, I think it does. But I'm cool with removing it since the behavior wouldn't be consistent.

File: rust/connlib/clients/android/src/lib.rs
Comment (thomaseizinger) [inline rust/connlib/clients/android/src/lib.rs:337]: What if we return the guard from here and store it within `Session`?

Comment (jamilbk) [reply rust/connlib/clients/android/src/lib.rs:337]: This is a good suggestion -- I'll take a stab at it this morning.

**Pull request merged**
